VideoWeb/VideoWeb/ClientApp/src/app/services/video-filter.service.ts

```diff
@@ -0,0 +1,185 @@
+            onFrame: async () => {
+                try {
+                    if (this.videoElement) {
+                        await this.selfieSegmentation.send({ image: this.videoElement });
+                    }
+                } catch (err) {
+                    this.logger.error(`${this.loggerPrefix} failed to send image to self segmentation mask`, err);
+                }
+            },
+            width: 1280,
+            height: 720
+        });
+        camera.start();
+    }
+
+    updateCameraStream(stream: MediaStream) {
+        this.videoElement.srcObject = stream;
+    }
+
+    startFilteredStream(): MediaStream {
+        this.canvasStream = this.canvasElement.captureStream();
+        return this.canvasStream;
+    }
+
+    updateFilter(filter: BackgroundFilter | null) {
+        this.logger.debug(`${this.loggerPrefix} Updating filter to ${filter}`);
+        if (filter) {
+            this.preferredFilterCache.set(filter);
+            this.activeFilter = filter;
+            this.filterOn = true;
+            this.logger.debug(`${this.loggerPrefix} Filter on`);
+            this._onFilterChanged.next(filter);
+        } else {
+            this.preferredFilterCache.clear();
+            this.activeFilter = null;
+            this.filterOn = false;
+            this.logger.debug(`${this.loggerPrefix} Filter off`);
+            this._onFilterChanged.next(null);
+        }
+    }
+
+    private onSelfieSegmentationResults(results: Results): void {
+        // Draw the overlays.
+        this.canvasCtx.save();
+        this.canvasCtx.clearRect(0, 0, this.canvasElement.width, this.canvasElement.height);
+
+        this.canvasCtx.drawImage(results.segmentationMask, 0, 0, this.canvasElement.width, this.canvasElement.height);
+
+        // Only overwrite existing pixels.
+        this.canvasCtx.globalCompositeOperation = 'source-in';
+
+        this.canvasCtx.drawImage(results.image, 0, 0, this.canvasElement.width, this.canvasElement.height);
+
+        // Only overwrite missing pixels.
+        this.canvasCtx.globalCompositeOperation = 'destination-atop';
+        this.applyEffect(results);
+        this.canvasCtx.restore();
+    }
+
+    private applyEffect(results: Results) {
+        switch (this.activeFilter) {
+            case BackgroundFilter.HMCTS:
+            case BackgroundFilter.SCTS:
+                this.applyVirtualBackgroundEffect();
+                break;
+            case BackgroundFilter.blur:
+                this.applyBlurEffect(results, true);
+                break;
+            default:
+                this.applyBlurEffect(results, false);
+                break;
+        }
+    }
+
+    private applyBlurEffect(results: Results, withBlur: boolean) {
+        if (withBlur) {
+            this.canvasCtx.filter = 'blur(50px)';
+        }
+        this.canvasCtx.drawImage(results.image, 0, 0, this.canvasElement.width, this.canvasElement.height);
+    }
```

Comment:
Maybe the name of the method could be clearer, feel it implies it will always have a blur.

Reply:
What would you suggest? The usage of the method is explicitly only when the active filter is blur